### ISSUE TITLE
YM2612 samples not plaing and YM2413 not supported.

### ISSUE THREAD

**Issue by system64MC** [libgme/game-music-emu#96]
Hi, I noticed 2 problems related to Game Music Emu.
    
The first one is samples not playing when playing some VGMs. I think supporting newer VGM features would solve the problem.
Another issue is YM2413 doesnt seem to be supported,  despite being added in VGM 1.10,  so some Master System VGMs will not play properly.
Here is a zip with 2 files that trigger the issue. The one named goddes2.vgm is for the YM2612 issue, the other named fantasyzone.vgm is for YM2413.
[VGMs.zip](https://github.com/user-attachments/files/16482783/VGMs.zip)

**Comment by Wohlstand** [comment]
Hello!
Right now VGM nodule supports only YM2612 and SN chips for a while. There are plans after 0.6.4 release to make an attempt to implement the rest of missing chips using the YMFM thing.

**Comment by system64MC** [comment]
> Hello! Right now VGM nodule supports only YM2612 and SN chips for a while. There are plans after 0.6.4 release to make an attempt to implement the rest of missing chips using the YMFM thing.

That would be great! It has so many chips such as OPL3, OPM, OPNA and many more. So maybe libgme would be able to support them.

**Comment by vitamin-caig** [comment]
And what about ym2413 support in kss? kode54's version supports it, but it's way too outdated for now...

**Comment by Wohlstand** [comment]
Hello! @kode54's was mostly his personal experiment, and he don't suggest to use it as mainstream, and instead, I polish this mainstream version. For now, it's planned to take the chipset from the "YMFM" library to bring the support for the rest of VGM-supported chips, and this one should be included too.

**Comment by vitamin-caig** [comment]
> Hi, isn't LibVGM licensed under GPL?

To be honest, I don't known and don't care about that, just using it in opensource project:)

**Comment by Wohlstand** [comment]
> PS. I guess, you may get rid of VGM support and cleanup code - there's mainstream libvgm for that:)

Actually, I tried to use libVGM for some of my experiments, and its functionality is not enough for my needs, mainly the ability to dynamically control temp, etc. It was done, but unfortunately, it works wrong, I wanted to fix that, but I became busy on other things, and I forgot about that.

**Comment by vitamin-caig** [comment]
> If you use that just for self, then okay

Yep, that's why I don't care. Your case may differ. So you may ask an author directly https://github.com/ValleyBell/libvgm

> its functionality is not enough for my needs, mainly the ability to dynamically control temp, etc

For my case, full support of devices (and good design especially comparing to awful VGMPlay) was the main point - there were way too many complaints about vgm playback via gme.

**Comment by Wohlstand** [comment]
> there were way too many complaints about vgm playback via gme

Just because historically nobody developed it well, and its support was so incomplete here. The libGME has major potential to support it better and wider, just need to take a time and effort. The same case with my libOPNMIDI: since my childhood I wanted to get a full-featured MIDI synthesizer with OPN2 chip after impressing from music in SEGA games and already having experience with OPL3 on my SB16 card. And, I just did libOPNMIDI by myself, after fiveteen years when I just got enough experience of programming.

**Comment by vitamin-caig** [comment]
> And, I just did libOPNMIDI by myself, after fiveteen years when I just got enough experience of programming.

I have to take a look:) I tried to support MIDI playback via mt32emu library about 10 years ago, the main problem was rendering performance on typical android devices.

**Comment by Wohlstand** [comment]
Anyway, I guess, I probably already suggested to implement MIDI with libADLMIDI and libOPNMIDI at zxtune at some moment, but nothing happen after that :thinking:

**Comment by freq-mod** [comment]
libADLMIDI is good, libOPNMIDI still has some features left to implement to be honest - PCM, PSG for example. Or at least ability to use Fluidsynth/TiMIDIty for PCM voices, but that's the player issue.

> Nuked OPN2 just don't work on phone, it requires even more resources to work.

it works on fairly recent (last few years) phones, it's not YM2608-LLE 😛 personally I wouldn't go below YMFM, all others are absurdly inaccurate

**Comment by Wohlstand** [comment]
> it works on fairly recent (last few years) phones

Even it works, it consumes too much of battery's energy. Imagine you run a stress test on all your CPU cores, and you'll eat phone's battery in the quickest way.
